nni/compression/pytorch/speedup/v2/dependency.py

```diff
@@ -0,0 +1,444 @@
+            continue
+        layers.extend(find_adjacent_layers(node, module, target_types, direction))
+    return layers
+
+
+def auto_set_denpendency_group_ids(graph_module: torch.fx.GraphModule,
+                                   config_list: List[Dict[str, Any]],
+                                   prune_type: str = 'Filter',
+                                   prune_axis: int = 1) -> List[Dict[str, Any]]:
+    """
+    Auto find the output dependency between all 'Conv2d', 'Linear', 'ConvTranspose2d', 'Embedding' modules,
+    then set the ``dependency_group_id`` in config list.
+
+    Note that a new dependency group id will be set as a shortcut in one config,
+    it will replace the old configured one in that config.
+
+    Parameters
+    ----------
+    graph_module : torch.fx.GraphModule
+        The traced model to be analyzed.
+    config_list : list
+        The compression config list.
+
+    Returns
+    config_list : list
+        The new config list with dependency group id.
+    """
+    def trans_target(target:str):
+        target = target.split('_')
+        return '.'.join(target)
+
+    channel_dependency = build_channel_dependency(graph_module, prune_type, prune_axis)
+    module2uid = {}
+    for d_set in channel_dependency:
+        uid = uuid4().hex
+        module2uid.update({trans_target(node.target): uid for node in d_set})
+
+    group_dependency = build_group_dependency(graph_module)
+    group_dependency = {trans_target(node.target): group_max for node, (group_max, group_min) in group_dependency.items()}
+
+    config_list = trans_legacy_config_list(config_list)
+    new_config_list = []
+    for config in config_list:
+        modules, public_config, _ = select_modules_by_config(graph_module, config)
+        for target in modules.keys():
+            sub_config = deepcopy(public_config)
+            if target in module2uid:
+                sub_config['dependency_group_id'] = module2uid[target]
+            if target in group_dependency:
+                sub_config['internal_metric_block'] = group_dependency[target]
+            new_config_list.append({
+                'op_names': [target],
+                **sub_config
+            })
+
+    return new_config_list
+
+
+def convert_dependency_to_set(dependency: Dict[Any, Set[Any]]) -> List[Set[Any]]:
+    """
+    Convert the dependency dict to sets of dependent nodes.
+
+    Parameters
+    ----------
+    dependency : Dict
+        The individual dependency mapping for the target graph.
+
+    Returns
+    -------
+    List
+        A list of dependencies for the target graph.
+    """
+    visited = set()
+    d_sets = []
+    for key, value in dependency.items():
+        if key not in visited:
+            tmp = set([key]) | value
+            visited.update(tmp)
+            if len(tmp) > 1:
+                d_sets.append(tmp)
+    return d_sets
+
+
+def build_weight_sharing_dependency(graph_module: torch.fx.GraphModule) -> List[List[torch.fx.Node]]:
```

Comment:
Could you please explain more about the weight-sharing dependency? What scenarios is it for? why it has such dependencies?

Reply:
this sounds like
```python
class Example(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.conv = nn.Conv2d(3, 3, 3, padding=1)
        self.conv2 = nn.Conv2d(3, 3, 3, padding=1)
        
    def forward(self, x):
        x = self.conv(x) + x
        x = self.conv2(x) + x
        x = self.conv(x) + x
        return x
```
where `self.conv` is repeatedly used in forward. in `torch.fx` we might prune ``self.conv`` twice.